crates/sp-lightclient/src/lib.rs

```diff
@@ -0,0 +1,308 @@
+        ImportError::DigestExtractionError(error)
+    }
+}
+
+/// Verifies and import headers.
+pub trait HeaderImporter<Header: HeaderT> {
+    /// Storage type to store headers and other computed details.
+    type Storage: Storage<Header>;
+
+    /// Verifies header, computes consensus values for block progress and stores the HeaderExt.
+    fn import_header(header: Header) -> Result<(), ImportError<HashOf<Header>>> {
+        // check if the header is already imported
+        match Self::Storage::header(header.hash()) {
+            Some(_) => Err(ImportError::HeaderAlreadyImported),
+            None => Ok(()),
+        }?;
+
+        // fetch parent header
+        let parent_header = Self::Storage::header(*header.parent_hash())
+            .ok_or_else(|| ImportError::MissingParent(header.hash()))?;
+
+        // TODO(ved): check for farmer equivocation
+
+        // verify global randomness, solution range, and salt from the parent header
+        let (global_randomness, solution_range, salt) =
+            verify_header_digest_with_parent(&parent_header, &header)?;
+
+        // extract subspace pre digest that contains the solution
+        let pre_digest = find_pre_digest(&header).map_err(|_| ImportError::InvalidPreDigest)?;
+
+        // slot must be strictly increasing from the parent header
+        verify_slot(&parent_header.header, &pre_digest)?;
+
+        // verify block signature
+        verify_block_signature(&header, &pre_digest.solution.public_key)?;
+
+        // verify solution
+        verify_solution(
+            &pre_digest.solution,
+            pre_digest.slot.into(),
+            VerifySolutionParams {
+                global_randomness: &global_randomness.global_randomness,
+                solution_range: solution_range.solution_range,
+                salt: salt.salt,
+                // TODO(ved): verify POAS once we have access to record root
+                piece_check_params: None,
+            },
+        )
+        .map_err(|_| ImportError::InvalidSolution)?;
+
+        let block_weight =
+            calculate_block_weight(&global_randomness.global_randomness, &pre_digest);
+        let total_weight = parent_header.total_weight + block_weight;
+
+        // last best header should ideally be parent header. if not check for forks and prune accordingly
+        let last_best_header = Self::Storage::best_header();
+        if last_best_header.header.hash() != parent_header.header.hash() {
+            let last_best_weight = last_best_header.total_weight;
+            if !match total_weight.cmp(&last_best_weight) {
+                // current weight is greater than last best. pick this header
+                Ordering::Greater => true,
+                // if weights are equal, pick the longest chain
+                Ordering::Equal => header.number() > last_best_header.header.number(),
+                // we already are on the best chain
+                Ordering::Less => false,
+            } {
```

Comment:
I'd store this in a variable before applying `!`, it is harder to read this way

Reply:
fair!